### PR TITLE
State history

### PR DESCRIPTION
**Pull request by wallbasher**
Added state history graph shown above existing graphs, these have full color state mapping specification.

Adds options to show hour bars every x hours

Somebody asked about doing this over on Reddit: https://www.reddit.com/r/homeassistant/comments/s74845/whats_the_proper_way_to_register_my/
https://community.home-assistant.io/t/history-graph-card-mixing-bars-and-graphs/214665

image

```
entities:
  - entity: sensor.hvac_temperature
    name: HVAC
    color: yellow
  - entity: sensor.basement_leak_temperature
    name: Basement
  - entity: sensor.smartthings_motion_temperature
    name: Upstairs
  - entity: sensor.openweathermap_temperature
    name: Outside
    color: purple
  - entity: sensor.basement_multisensor_temperature
    name: Johns Office
  - entity: sensor.upstairs_multisensor_temperature_2
    name: Master Bedroom
  - entity: sensor.hvac_action
    name: HVAC Action
    style: historyBar
    states:
      - value: idle
        label: 'Off'
        color: black
        opacity: 0.1
      - value: heating
        color: '#B71C1C'
        label: Heat
      - value: cooling
        label: Cool
        color: '#1A237E'
      - value: fan
        color: '#1565C0'
        label: Fan
      - value: unknown
        label: Unavailable
        color: '#9E9E9E'
hours_to_show: 48
line_width: 3
points_per_hour: 2
refresh_interval: 0
show:
  fill: false
  labeles_secondary: true
  labels: true
show_legend: false
show_line: false
show_points: false
historyGraph:
  bar_height: 20
  bar_gap: 10
y_axis: secondary
type: custom:mini-graph-card

```

image

```
entities:
  - entity: person.john_weilbacher
    style: historyBar
    show_legend: false
    name: John
    states:
      - value: home
        label: Home
        color: '#1b1b1b'
        opacity: 0.95
      - value: Target
        color: '#B71C1C'
        label: Target
      - value: not_home
        label: Out
        color: '#1A237E'
      - value: The Valley
        color: '#1565C0'
        label: The Valley
      - value: Waterloo
        color: '#D84315'
        label: Waterloo
      - value: Carly's
        label: Carly's
        color: '#FF9100'
  - entity: person.lyndsy_weilbacher
    style: historyBar
    name: Lyndsy
    show_legend: false
    states:
      - value: home
        label: Home
        color: '#1b1b1b'
        opacity: 0.95
      - value: Target
        color: '#B71C1C'
        label: Target
      - value: not_home
        label: Out
        color: '#1A237E'
      - value: The Valley
        color: '#1565C0'
        label: The Valley
      - value: Waterloo
        color: '#D84315'
        label: Waterloo
      - value: Carly's
        label: Carly's
        color: '#FF9100'
      - value: unavailable
        label: Unavailable
        color: '#9E9E9E'
  - entity: lock.front_door
    style: historyBar
    name: Front Door
    show_legend: false
    states:
      - value: unlocked
        label: Unlocked
        color: '#880E4F'
      - value: locked
        label: Locked
        color: '#1b1b1b'
        opacity: 0.95
      - value: unavailable
        label: Unavailable
        color: '#9E9E9E'
  - entity: lock.back_door
    style: historyBar
    name: Back Door
    show_legend: false
    states:
      - value: unlocked
        label: Unlocked
        color: '#880E4F'
      - value: locked
        label: Locked
        color: '#1b1b1b'
        opacity: 0.95
      - value: unavailable
        label: Unavailable
        color: '#9E9E9E'
  - entity: cover.garage_door
    style: historyBar
    name: Garage
    show_legend: false
    states:
      - value: open
        label: Open
        color: '#880E4F'
      - value: closing
        label: Closing
        color: '#880E4F'
      - value: unavailable
        label: Unavailable
        color: '#9E9E9E'
      - value: opening
        label: Opening
        color: '#880E4F'
      - value: closed
        label: Closed
        color: '#1b1b1b'
        opacity: 0.95
  - entity: binary_sensor.basement_leak_moisture
    style: historyBar
    name: Basement Drain
    states:
      - value: 'off'
        label: Dry
        color: '#1b1b1b'
        opacity: 0.95
      - value: 'on'
        label: Leak
        color: '#880E4F'
      - value: unavailable
        label: Unavailable
        color: '#880E4F'
  - entity: binary_sensor.front_door_motion
    style: historyBar
    name: Front Door Motion
    states:
      - value: 'off'
        label: None
        color: '#1b1b1b'
        opacity: 0.95
      - value: 'on'
        label: Motion
        color: '#880E4F'
  - entity: binary_sensor.front_door_ding
    style: historyBar
    name: Doorbell
    states:
      - value: 'off'
        label: None
        color: '#1b1b1b'
        opacity: 0.95
      - value: 'on'
        label: Ding
        color: '#880E4F'
historyGraph:
  bar_height: 15
  bar_gap: 5
hours_to_show: 12
line_width: 3
points_per_hour: 12
refresh_interval: 5
lines_every_x_hour: 1
height: 500
fill: false
type: custom:mini-graph-card
```

image
```
entities:
  - entity: binary_sensor.ac
    style: historyBar
    name: AC
  - entity: binary_sensor.coffee_maker
    style: historyBar
    name: Coffee Maker
  - entity: binary_sensor.dishwasher
    style: historyBar
    name: Dishwasher
  - entity: binary_sensor.dryer
    style: historyBar
    name: Dryer
  - entity: binary_sensor.electric_blanket
    style: historyBar
    name: Heated Blanket
  - entity: binary_sensor.furnace_2
    style: historyBar
    name: Furnace
  - entity: binary_sensor.heat_3
    style: historyBar
    name: Heat
  - entity: binary_sensor.heat_5
    style: historyBar
    name: Heat
  - entity: binary_sensor.heat_6
    style: historyBar
    name: Heat
  - entity: binary_sensor.heat_7
    style: historyBar
    name: Heat
  - entity: binary_sensor.heat_8
    style: historyBar
    name: Heat
  - entity: binary_sensor.hot_water_heater_fan
    style: historyBar
    name: Hot Water Fan
  - entity: binary_sensor.microwave
    style: historyBar
    name: Microwave
  - entity: binary_sensor.oven
    style: historyBar
    name: Oven
  - entity: binary_sensor.stove
    style: historyBar
    name: Stove
  - entity: binary_sensor.washer
    style: historyBar
    name: Washer
  - entity: sensor.hvac_action
    name: HVAC Action
    style: historyBar
  - entity: sensor.energy_usage
    name: Usage
hours_to_show: 120
line_width: 3
points_per_hour: 0.5
refresh_interval: 0
show:
  fill: false
show_legend: false
show_line: true
show_points: false
historyGraph:
  bar_height: 10
  bar_gap: 2
y_axis: secondary
states:
  - value: 'off'
    label: 'Off'
    color: '#1b1b1b'
    opacity: 0.95
  - value: 'on'
    color: '#B71C1C'
    label: 'On'
  - value: idle
    label: 'Off'
    color: '#1b1b1b'
    opacity: 0.95
  - value: heating
    color: '#B71C1C'
    label: Heat
  - value: cooling
    label: Cool
    color: '#1A237E'
  - value: fan
    color: '#1565C0'
    label: Fan
  - value: unknown
    label: Unavailable
    color: '#9E9E9E'
type: custom:mini-graph-card
```

### CODE REVIEW

File: src/style.js [hunk 1]
```diff
@@ -233,6 +233,7 @@ const style = css`
     flex-direction: column;
     margin-top: auto;
     width: 100%;
+    padding: 0px;
```
Comment (wallbasher) [inline src/style.js:236]: Make this a value that can be set, for some reason when I set this here to another value the history bars on top would not show up.  This is the only "impact" on the original library.

**Review by jlsjonas — CHANGES_REQUESTED** [review]
Hi, thanks for your contribution!
I'm not entirely convinced on the added complexity & duplicated code to accomplish this though, but I'm not against the concept of adding it.

We plan on reworking the configuration & general codebase to make it more structured, it might benefit to wait for that release to implement this feature in a more sustainable way.

(only partially reviewed, for this reason; apologies for it getting stuck in pending for a while)

File: src/main.js [hunk 2]
```diff
@@ -251,12 +287,23 @@ class MiniGraphCard extends LitElement {
...
+          <div class="state">
+            <span class="state__value ellipsis" style=${color}>
+              ${this.computeState(state)}
+            </span>
+            <span class="state__uom ellipsis" style=${color}>
+              ${this.computeUom(entity || 0)}
+            </span>
+            ${this.renderStateTime()}
+          </div>
+          <div class="states--secondary">${this.normalEntities().map((ent, i) => this.renderState(ent, i))}</div>
```
[the remark this answers is not in the document]
Comment (wallbasher) [reply src/main.js:309]: Sorry about that must have missed it, I started this a year ago then finally got around to finishing it.

**Comment by wallbasher** [comment]
I really like the mini-graph, but if falls short in having States displayed.  I wanted something that could get close to what exists in the ecobee system.  

Image examples:
https://www.reddit.com/r/ecobee/comments/gh6kli/help_me_understand_the_system_usage_graph/

**Pull request closed without merging**